### ISSUE TITLE
Addition operator is interpreted as subtraction operator if called inside a for-loop and combined with multiplication or division operator

### ISSUE THREAD

**Issue by stephan-gruen** [dart-lang/sdk#37337]
Addition operator is interpreted as subtraction operator if called inside a for-loop and combined with multiplication or division operator.

The behaviour is like expected with serve and build in and prior dart-sdk 2.2.1-dev.1.0

The behaviour is like expected with 
"D:\dart-sdk\dart-sdk 2.4.0-dev.0.0\bin\pub.bat" global run webdev serve web:53322
"D:\dart-sdk\dart-sdk 2.5.0-dev.0.0\bin\pub.bat" global run webdev serve web:53322

But unexpected in the actual build with
"D:\dart-sdk\dart-sdk 2.4.0-dev.0.0\bin\pub.bat" global run webdev build --output=web:build
"D:\dart-sdk\dart-sdk 2.5.0-dev.0.0\bin\pub.bat" global run webdev build --output=web:build

Normal behaviour outside a for-loop OR if the value is set to 100 INSIDE the loop:
```
v = 100;
v += 50;   // expected: 150
150

v = 100;
v += v * 0.5;   // expected: 150
150

v = 100;
v = v + (v * 0.5);   // expected: 150
150

v = 100;
v += v / 2;   // expected: 150
150

v = 100;
v = v + (v / 2);   // expected: 150
150
```

Wrong behaviour inside a for-loop if the value is set outsite the loop.
Addition operator is interpreted as subtraction operator if called inside a for-loop and combined with multiplication or division operator.
```
v = 100;
for (var i=0; i<3; i++) ...

v += 50;   // expected: 150, 200, 250
150
200
250

v += v * 0.5;   // expected: 150, 225, 337.5
50
25
12.5

v = v + (v * 0.5);   // expected: 150, 225, 337.5
50
25
12.5

v += v / 2;   // expected: 150, 225, 337.5
50
25
12.5

v = v + (v / 2);   // expected: 150, 225, 337.5
50
25
12.5
```

Source:
```
   double v;
    
    v = 100.0;
    v += 50;
    print('v = 100;');
    print('v += 50;   // expected: 150');
    print(v);
    print('');

    v = 100.0;
    v += v * 0.5;
    print('v = 100;');
    print('v += v * 0.5;   // expected: 150');
    print(v);
    print('');

    v = 100.0;
    v = v + (v * 0.5);
    print('v = 100;');
    print('v = v + (v * 0.5);   // expected: 150');
    print(v);
    print('');

    v = 100.0;
    v += v / 2;
    print('v = 100;');
    print('v += v / 2;   // expected: 150');
    print(v);
    print('');

    v = 100.0;
    v = v + (v / 2);
    print('v = 100;');
    print('v = v + (v / 2);   // expected: 150');
    print(v);
    print('');

    print('Inside a for-loop, but the value is preset outsite the loop:');
    print('v += 50;   // expected: 150, 200, 250');
    v = 100.0;
    for (var i=0; i<3; i++) {
      v += 50;
      print(v);
    }
    print('');

    print('v += v * 0.5;   // expected: 150, 225, 337.5');
    v = 100.0;
    for (var i=0; i<3; i++) {
      v += v * 0.5;
      print(v);
    }
    print('');

    print('v = v + (v * 0.5);   // expected: 150, 225, 337.5');
    v = 100.0;
    for (var i=0; i<3; i++) {
      v = v + (v * 0.5);
      print(v);
    }
    print('');

    print('v += v / 2;   // expected: 150, 225, 337.5');
    v = 100.0;
    for (var i=0; i<3; i++) {
      v += v / 2;
      print(v);
    }
    print('');

    print('v = v + (v / 2);   // expected: 150, 225, 337.5');
    v = 100.0;
    for (var i=0; i<3; i++) {
      v = v + (v / 2);
      print(v);
    }
```

**Comment by lrhn** [comment]
Cant know for sure whether `+=` acts as `-=` or just `=`, but it's definitely wrong.

Doesn't reproduce on current VM or dart2js. Is this a DDC issue?

**Comment by stephan-gruen** [comment]
Even Stranger Things

In search for a workaround I found this out:

```
"D:\dart-sdk\dart-sdk 2.5.0-dev.0.0\bin\pub.bat" global run webdev build --output=web:build
[INFO] Reading cached asset graph completed, took 870ms
[INFO] Checking for updates since last build completed, took 1.4s
[INFO]build_web_compilers:entrypoint on web/main.dart: Running dart2js with --minify --packages=.package-32fc821bdb19f486eb3a3d7b257ac9e4 -oweb/main.dart.js web/main.dart
```

```
double a;
double b;
double v;
double w;
```

Not a workaround:
```
v = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  v = v + a;
  print('v= $v');
}
```
Output:
```
v= 50
v= 25
v= 12.5
```

This worked!
```
v = 100.0;
w = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  v = v + a;
  w = w + a;  // a!! This line corrected v!
  print('v= $v  w= $w');
}
```
Output:
```
v= 150  w= 150
v= 225  w= 225
v= 337.5  w= 337.5
```

Back to the old behavior:
```
v = 100.0;
w = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  b = w * 0.5;
  v = v + a;
  w = w + b;  // b!!
  print('v= $v  w= $w');

}
```
Output:
```
v= 50  w= 50
v= 25  w= 25
v= 12.5  w= 12.5
```

**Comment by stephan-gruen** [comment]
Follow up:

All it needs is to call the var that contains the product (a or b) once to trigger a correct output:

```
v = 100.0;
w = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  b = w * 0.5;
  v = v + a;
  w = w + b;
  print(a);  // <--- Trigger for v
  print('v=$v  w=$w');
}
```

```
50
v=150  w=50
75
v=225  w=25
112.5
v=337.5  w=12.5
```


```
v = 100.0;
w = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  b = w * 0.5;
  v = v + a;
  w = w + b;
  print(b);  // <--- Trigger for w
  print('v=$v  w=$w');
}
```

Output:
```
50
v=50  w=150
75
v=25  w=225
112.5
v=12.5  w=337.5
```

**Comment by stephan-gruen** [comment]
```
v = 100.0;
w = 100.0;
for (var i=0; i<3; i++) {
  a = v * 0.5;
  b = w * 0.5;
  v = v + a;
  w = w + b;
  print(a);
  print('v=$v  w=$w');
}
```

Output:
```
50
v=150  w=50
75
v=225  w=25
112.5
v=337.5  w=12.5
```


The minified js code from main.dart.js:

```
for(u=100,q=100,t=0;t<3;++t){
	p=u*0.5
	u+=p
	q*=0.5
	
	s=H.i(p)
	r=$.b7
	if(r==null)
		H.b1(s)
	else 
		r.$1(s)
	
	s="v="+H.i(u)+"  w="+H.i(q)
	r=$.b7
	if(r==null)
		H.b1(s)
	else 
		r.$1(s)}
```

**Comment by rakudrama** [comment]
This _complete input_ shows the problem
```
main() {
  double v = 0, w = 0;
  double a, b;
  v = 100.0;
  w = 100.0;
  for (var i=0; i<3; i++) {
    a = v * 0.5;
    b = w * 0.5;
    v = v + a;
    w = w + b;
    print(a);
    print('v=$v  w=$w');
  }
}
```
The `-O1` output shows what appears to be some confusion between `w` and `b`, possibly during variable allocation:
```
    main: function() {
      var v, w, i, a;
      for (v = 100, w = 100, i = 0; i < 3; ++i) {
        a = v * 0.5;
        v += a;
        w *= 0.5;
        H.printString(H.S(a));
        H.printString("v=" + H.S(v) + "  w=" + H.S(w));
      }
    }
```

**Comment by rakudrama** [comment]
Fixed by c7524cc29613caa00326c5cb7c5d78a4ecec9618
@stephan-gruen Thanks for all your investigating!
